gateway-ha/src/main/java/io/trino/gateway/ha/router/ProcessedRequest.java

```diff
@@ -0,0 +1,524 @@
+            // Note that this does necessarily not match the Resource Group query type. io.trino.util.StatementUtils maps statement
+            // types to a smaller list of query types.
+            getTables(statement, tables);
+
+            if (catalogs.isEmpty()) {
+                catalogs = tables.stream().map(q -> q.getParts().getFirst()).collect(Collectors.toSet());
+            }
+            if (schemas.isEmpty()) {
+                schemas = tables.stream().map(q -> q.getParts().get(1)).collect(Collectors.toSet());
+            }
+            catalogSchemas = tables.stream()
+                    .map(qualifiedName -> String.format("%s.%s", qualifiedName.getParts().getFirst(), qualifiedName.getParts().get(1)))
+                    .collect(Collectors.toSet());
+        }
+        catch (IOException e) {
+            log.warn("Error extracting request body for rules processing: " + e.getMessage());
+        }
+        catch (ParsingException e) {
+            log.info("Could not parse request body as SQL: " + body + "; Message: " + e.getMessage());
+        }
+    }
+
+    private Map<String, String> getPreparedStatements(HttpServletRequest request)
+    {
+        Map<String, String> preparedStatementsMap = new HashMap<>();
+        Enumeration<String> headers = request.getHeaders(TRINO_PREPARED_STATEMENT_HEADER_NAME);
+        if (headers == null) {
+            return preparedStatementsMap;
+        }
+        while (headers.hasMoreElements()) {
+            String[] preparedStatementsArray = headers.nextElement().split(",");
+            for (String preparedStatement : preparedStatementsArray) {
+                String[] nameValue = preparedStatement.split("=");
+                if (nameValue.length != 2) {
+                    throw new InputMismatchException(String.format("preparedStatement must be formatted as name=value, but is %s", preparedStatement));
+                }
+                preparedStatementsMap.put(URLDecoder.decode(nameValue[0], UTF_8), URLDecoder.decode(decodePreparedStatementFromHeader(nameValue[1]), UTF_8));
+            }
+        }
+        return preparedStatementsMap;
+    }
+
+    public String decodePreparedStatementFromHeader(String headerValue)
+    {
+        // From io.trino.server.protocol.PreparedStatementEncoder
+        String prefix = "$zstd:";
+        if (!headerValue.startsWith(prefix)) {
+            return headerValue;
+        }
+
+        String encoded = headerValue.substring(prefix.length());
+        byte[] compressed = base64Url().decode(encoded);
+
+        byte[] preparedStatement = new byte[toIntExact(ZstdDecompressor.getDecompressedSize(compressed, 0, compressed.length))];
+        new ZstdDecompressor().decompress(compressed, 0, compressed.length, preparedStatement, 0, preparedStatement.length);
+        return new String(preparedStatement, UTF_8);
+    }
+
+    private void getTables(Node node, Set<QualifiedName> tables)
```

Comment:
Please add a unit test for this method.

Reply:
This method is tested by `TestRoutingGroupSelector.testRequestProcessorQueryDetails`. For the SQL
```
SELECT x.*, y.*, z.* FROM catx.schemx.tblx a, schemy.tbly y, tblz z
```
The matched rule checks
```
processedRequest.tablesContains("cat_default.\"schem_\\\"default\".tblz")
  && processedRequest.tablesContains("cat_default.schemy.tbly")
  && processedRequest.tablesContains("catx.schemx.tblx")
  && processedRequest.getSchemas().contains("schemy")
  && processedRequest.getCatalogs().contains("catx")
```

What would you like to see in a unit test?